README

 Fill this in.


Changes
Revision history for Perl extension Alchemy::Blog.

-[ Problems To Ponder ]------------------------------------------------

	These are tenative.

	Should be able to use location /, and have it all work.
		- Just to be fun ;)

	/blog 
		/main						# Show the default blog.
			/PAGING					# Allow paging on the frontpage.

		/author						# List all of the authors.
			/AUTHOR					# List a specfic author.

		/calendar
				/YEAR				# Show a years archive
				/YEAR-MONTH			# Show a months archive
				/YEAR-MONTH-DAY		# Show a days archive
				
		/label
			/LABEL(s)				# Show arhive for a category
				/PAGING				# Allow for paging.

		/entry					
			/IDENT					# Show a particular entry.
				

		/comment					# Show recent comments.
				/IDENT				# Show article comments
			/post
				/IDENT				# Post a comment

	/admin
			Blog edit
			Authors
			Article Types
			Notification
			Entrys & Comments

-[ Todo ]--------------------------------------------------------------


 	- A real comment implementation.
		- Inline or not comments option. ( article override )
		- Add open comment period. 0 for forever. ( article override,
		  default in the blog/category )
		- Comments inline on a story or not by article option.
		- editable comments, should note that they have been edited in
		  the public view.
	- Multiple Authors, and the ability to have different users not able
	  to edit each others entries.
		- Authors may be limited to a particular blog.
		- Pick users to admin a blog... 
		- Assign authors to a particular blog.
 	- Better RSS support.
 	- File management for entries.
	- Entry types to allow for multiple entry formats, think photo blog
	  vs regular blog vs just link lists.
		- How do we do blogmarks ? Is it a category option on what to show?
		- How do I do a photo book ?
 	- Simpler interface for posting entries.
 	- Built in support for multiple ping hosts.
	- Multiple archive views: A list, like the current incarnation, and
	  an actual calendar.
 	- Automatic thumbnail generation.
	- Entry validation, and automatic entry encoding, to simplify
	  posting for novices.
	- *Possible* Plug-ins, to allow for stuff, this likely will be a
	  late addition. 
	- Ident length may be set by blog.
	- Add the help documentation.
	- Allow prev/next by article, like a photoblog does.
	
	- Add the ability to manage files and generate links inline.

	- Add the help documentation.

-[ Revision History ]--------------------------------------------------

Figure out why the file upload isn't working.

0.43 Fri Sep  2 08:42:02 CDT 2005
	- Updating to the Apache2::* naming convention for mod_perl 2.000001

0.42 Sun Mar 20 14:31:53 CST 2005
	- Added a span around the date for the blog_roll.pl

0.41 Thu Feb 17 20:49:35 CST 2005
	- Updated the schema for the rewrite. [ns]
	- Moved to using $site->_decline over returning Apache::DECLINED [ns]
	- Added blog_amwish to the source tree [pe]

0.40 Thu Dec 23 21:25:10 CST 2004
	- Ported to mod_perl2
	- moved Blog.conf to httpd.conf
	- moved schema.
	- Everything seems to work with mod_perl2, need to do the file
	  upload correctly once an answer is found.

0.35 Tue Aug 10 20:43:49 CDT 2004
	- Fixed another bug in comment deletion.

0.34 Tue Aug 10 08:21:26 CDT 2004
	- Fixed time formating bug in the blog roll.

0.33 Sat Jul 31 21:25:37 CDT 2004
	- Moved to using ht_checkbox.
	- Added the where clause for sql_update.

0.32 Thu Jul 22 21:33:47 CDT 2004
	- Started on the authors to users link.
	- Fixed an archive bug.
	- Adjusted Blog_Post_Days to default to always open, rather than 7
	  days. This really needs to be put into the database.

0.31 Sat Jul 17 22:03:40 CDT 2004
	- Added the read/set for coookie comment posters.
	- Fixed a bug in the counting resulting from a comment being
	  removed, must use the *correct* id when selecting from the
	  database.
	- Flushed out the view unapproved comments page a bit.
	- Added the email view to all admin comments.
	- Fixed a problem in the comment classing in the admin pages.
	- Fixed the classing on comment form for the admin section.
	- Add recent comments page and a new setvar to limit it.
	  Blog_Recent_Show, which defaults to 10 comments. 
	- Put the article title and a link back to the original article  on
	  the seperate comments page. 
	- Added comments / read more to the archive pages. Days got the
	  comments links and their titles linked which the months got the
	  "Read Full Article" treatment.

0.30 Fri Jul 16 08:43:20 CDT 2004
	- Fixed article count math. 
	- fix where admin comments pages redirect too, they now go back to
	  the particular article view of the comments.
	- Fixed the bug that doesn't catch www.domain.com in sanitize_text
	- Fixed the textareas to have sizes for comments.

0.29 Wed Jul 14 21:01:20 CDT 2004
	- Fixed a bug in the article posting.
	- Fixed a bug in the force_auth.

0.28 Tue Jul 13 22:11:03 CDT 2004
	- Adjusted the archive to show all years, rather than just the
	  current year only.
	- Stubbed out the Comment admin module.
	- Moved do_rss from Blog::Viewer to Blog::Viewer::Syndication.do_main
	- Removed created from the bg_article table, it is not needed.
	- Removed ident from bg_article_type, it is not needed.
	- Updated the MANIFEST for the new files.
	- Added auto-generated idents for articles.
	- Adjusted internal datefmt, timefmt, and datetimefmt variables to
	  match the new keys.
	- Adjusted the internal variables used for the various module roots.
	  Added a couple of new ones to handle the new modules.
	- Added $site->section_info( $section_ident ) to grab section
	  information.
	- Added $site->author_info( $author_id ) to grab the author info.
	- Moved the main, ie full, article view out of do_archive and into
	  do_article. This can be fixed for existing URLs by using the
	  following mod_rewrite rule:
	  	RewriteEngine On
	  	RewriteRule ^/blog/archive/(.*)/(.*)/(.*) /blog/article/$1/$2/$3 [R]
	- Updated all of the internal links to deal with the archive ->

lib/Alchemy/Blog/Article/Author.pm
package Alchemy::Blog::Article::Author;
######################################################################
# $Id: Author.pm,v 1.11 2005/09/02 14:05:29 nstudt Exp $
# $Date: 2005/09/02 14:05:29 $
######################################################################
use strict;
use Apache2::Request;
use Apache2::Upload;
use APR::Const -compile => qw(:common);

use KrKit::Control;
use KrKit::DB;
use KrKit::Handler;
use KrKit::HTML qw( :all );
use KrKit::SQL;
use KrKit::Validate;

use Alchemy::Blog;

use vars qw( @ISA );

############################################################
# Variables                                                #
############################################################
@ISA = ( 'Alchemy::Blog', 'KrKit::Handler' );

############################################################
# Functions                                                #
############################################################

#-------------------------------------------------
# author_checkvals( $site, $in, $id )
#-------------------------------------------------
sub author_checkvals {
	my ( $site, $in, $id ) = @_;

	my @errors;

	$id = 0 if ( ! is_integer( $id ) );

	if ( ! is_text( $in->{name} ) ) {
		push( @errors, 'Enter a name for this author.'. ht_br() );
	}

	if ( ! is_email( $in->{email} ) ) {
		push( @errors, 'Enter an e-mail address for this author.'.  ht_br() );
	}

	if ( ! is_ident( $in->{ident} ) ) {
		push( @errors, 'Enter an ident for this author.'. ht_br() );
	}
	else {
		my $sth = db_query( $$site{dbh}, 'see if used',
							'SELECT count(id) FROM bg_authors ',
							'WHERE ident = ', sql_str( $in->{ident} ) );

		my ( $count ) = db_next( $sth );

		db_finish( $sth );

		if ( $id && $in->{oident} ne $in->{ident} ) {
			push( @errors, 'Ident already in use.'. ht_br() );
		}

		if ( ! $id && $count ) {
			push( @errors, 'Ident already in use.'. ht_br() );
		}
	}

	return( @errors );
} # END author_checkvals( $in )

#-------------------------------------------------
# author_form( $site, $in )
#-------------------------------------------------
sub author_form {
	my ( $site, $in ) = @_;

	return( ht_form_js( $$site{uri} ),	
			ht_div( { 'class' => 'box' } ),
			ht_table( {} ),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'Name',
					ht_help( $$site{help}, 'item', 'a:bg:aa:name' ) ),
			ht_td( 	{ 'class' => 'dta' },
					ht_input( 'name', 'text', $in, 'size="30"' ) ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'E-mail',
					ht_help( $$site{help}, 'item', 'a:bg:aa:email' ) ),
			ht_td( 	{ 'class' => 'dta' },
					ht_input( 'email', 'text', $in, 'size="30"' ) ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'Ident',
					ht_help( $$site{help}, 'item', 'a:bg:aa:name' ) ),
			ht_td( 	{ 'class' => 'dta' }, 
					ht_input( 'oident', 'hidden', $in ),
					ht_input( 'ident', 'text', $in ) ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'Bio',
					ht_help( $$site{help}, 'item', 'a:bg:aa:bio' ) ),
			ht_td( 	{ 'class' => 'dta' }, 
					ht_input( 	'bio', 'textarea', $in, 
								'cols="40" rows="10"' ) ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'colspan' => '2', 'class' => 'rshd' }, 
					ht_submit( 'submit', 'Save' ),
					ht_submit( 'cancel', 'Cancel' ) ),
			ht_utr(),

			ht_utable(),
			ht_udiv(),
			ht_uform() );
} # END author_form

#-------------------------------------------------
# $site->do_add( $r )
#-------------------------------------------------
sub do_add {
	my ( $site, $r ) = @_;

	my $in 				= $site->param( Apache2::Request->new( $r ) );
	$$site{page_title}	.= 'Add';
	
	return( $site->_relocate( $r, $$site{rootp} ) ) if ( $in->{cancel} );

	if ( ! ( my @errors = author_checkvals( $site, $in ) ) ) {

		$in->{bio} =~ s/(\r\n)|\r/\n/g;

		db_run( $$site{dbh}, 'insert new author',
				sql_insert( 'bg_authors',
							'ident' 	=> sql_str( $in->{ident} ), 
							'name' 		=> sql_str( $in->{name} ), 
							'email' 	=> sql_str( $in->{email} ), 
							'bio'		=> sql_str( $in->{bio} ) ) );
		
		my $id = db_lastseq( $$site{dbh}, 'bg_authors_seq' );

		db_commit( $$site{dbh} );

		return( $site->_relocate( $r, "$$site{rootp}/view/$id" ) );
	}
	else {
		if ( $r->method eq 'POST' ) {
			return( @errors, author_form( $site, $in ) );
		}
		else {
			return( author_form( $site, $in ) );
		}
	}
} # END $site->do_add

#-------------------------------------------------
# $site->do_delete( $r, $id, $yes )
#-------------------------------------------------
sub do_delete {
	my ( $site, $r, $id, $yes ) = @_;

	my $in 				= $site->param( Apache2::Request->new( $r ) );
	$$site{page_title} 	.= 'Delete';

	return( 'Invalid id.' ) 						if ( ! is_integer( $id ) );

	if ( defined $in->{cancel} ) {
		return( $site->_relocate( $r, $$site{rootp} ) );
	}

	my $ath = db_query( $$site{dbh}, 'see if used',
						'SELECT count(id) FROM bg_articles ',
						'WHERE bg_author_id = ', sql_num( $id ) );
	 
	my ( $count ) = db_next( $ath );

	db_finish( $ath );

	if ( $count ) {
		return( ht_div( { 'class' => 'box' } ),
				ht_table( {} ),
				ht_tr(),
					ht_td( 	{ 'class' => 'dta' }, 
							q!Unable to delete, still in use.! ),
				ht_utr(),
				ht_tr(),
					ht_td( { 'class' => 'rshd' }, 
							ht_a( $$site{rootp}, 'Back to Listing' ) ),
				ht_utr(),
				ht_utable(),
				ht_udiv() );
	}

	if ( ( defined $yes ) && ( $yes eq 'yes' ) ) {

		my $sth = db_query( $$site{dbh}, 'get photo',
							'SELECT photo FROM bg_authors WHERE id = ',
							sql_num( $id ) );

		my ( $photo ) = db_next( $sth );

		db_finish( $sth );

		if ( is_text( $photo ) ) {
			unlink( "$$site{file_path}/$photo" ) 
				or die "Can't remove photo: $!";
		}

		db_run( $$site{dbh}, 'remove the role',
				'DELETE FROM bg_authors WHERE id = ', sql_num( $id ) );

		db_commit( $$site{dbh} );

		return( $site->_relocate( $r, $$site{rootp} ) );
	}
	else {
		# Look up the role information.
		my $sth = db_query( $$site{dbh}, 'get role information',
							'SELECT name FROM bg_authors WHERE id = ',
							sql_num( $id ) );

		my ( $name ) = db_next( $sth );

		db_finish( $sth );

		return( ht_form_js( "$$site{uri}/yes" ), 
				ht_div( { 'class' => 'box' } ),
				ht_table( {} ),
				ht_tr(),
					ht_td( 	{ 'class' => 'dta' }, 
							qq!Delete the author "$name"?! ),
				ht_utr(),
				ht_tr(),
					ht_td( { 'class' => 'rshd' }, 
							ht_submit( 'submit', 'Continue with Delete' ),
							ht_submit( 'cancel', 'Cancel' ) ),
				ht_utr(),
				ht_utable(),
				ht_udiv(),
				ht_uform() );
	}
} # END $site->do_delete

#-------------------------------------------------
# $site->do_edit( $r, $id )
#-------------------------------------------------
sub do_edit {
	my ( $site, $r, $id ) = @_;

	my $in 				= $site->param( Apache2::Request->new( $r ) );
	$$site{page_title}	.= 'Update';

	return( 'Invalid id.' ) if ( ! is_integer( $id ) );

	if ( $in->{cancel} ) {
		return( $site->_relocate( $r, $$site{rootp} ) );
	}

	if ( ! ( my @errors = author_checkvals( $site, $in, $id ) ) ) {

		$in->{bio} =~ s/(\r\n)|\r/\n/g;

		db_run( $$site{dbh}, 'insert new author',
				sql_update( 'bg_authors', 'WHERE id = '. sql_num( $id ),
							'ident' 	=> sql_str( $in->{ident} ), 
							'name' 		=> sql_str( $in->{name} ), 
							'email' 	=> sql_str( $in->{email} ), 
							'bio'		=> sql_str( $in->{bio} ) ) );

		db_commit( $$site{dbh} );

		# Send to the photo page. ? or the view ?
		return( $site->_relocate( $r, "$$site{rootp}/view/$id" ) );
	}
	else {
		my $sth = db_query( $$site{dbh}, 'get old values',
							'SELECT ident, name, email, bio ',
							'FROM bg_authors WHERE id = ', sql_num( $id ) );

		while ( my ( $ident, $name, $email, $bio ) = db_next( $sth ) ) {
			$in->{oident} 	= $ident;
			$in->{ident}	= $ident 	if ( ! defined $in->{ident} );
			$in->{name} 	= $name 	if ( ! defined $in->{name} );
			$in->{email}	= $email 	if ( ! defined $in->{email} );
			$in->{bio} 		= $bio 		if ( ! defined $in->{bio} );
		}

		db_finish( $sth );

		if ( $r->method eq 'POST' ) {
			return( @errors, author_form( $site, $in ) );
		}
		else {
			return( author_form( $site, $in ) );
		}
	}
} # END $site->do_edit

#-------------------------------------------------
# $site->do_main( $r )
#-------------------------------------------------
sub do_main {
	my ( $site, $r ) = @_;

	$$site{page_title} .= 'Listing';

	my @lines = ( 	ht_div( { 'class' => 'box' } ),
					ht_table(),

					ht_tr(),
					ht_td( 	{ 'class' => 'shd' }, 'Author' ),
					ht_td( 	{ 'class' => 'shd' }, 'Ident' ),
					ht_td( 	{ 'class' => 'rshd' }, 
							'[', ht_a( "$$site{rootp}/add", 'Add' ), '|',
							 ht_a( $$site{root_article}, 'Articles' ), ']', ),
					ht_utr() );

	my $sth = db_query( $$site{dbh}, 'list all authors',
						'SELECT id, ident, name FROM bg_authors ',
						'ORDER BY name' );
	
	while ( my ( $id, $ident, $name ) = db_next( $sth ) ) {

		push( @lines, 	ht_tr(),
						ht_td( 	{ 'class' => 'dta' }, 
								ht_a( "$$site{rootp}/view/$id", $name ) ),
						ht_td( 	{ 'class' => 'dta' }, $ident ),
						ht_td( 	{ 'class' => 'rdta' }, 
								'[', 
								ht_a( "$$site{rootp}/edit/$id", 'Edit' ),  '|',
								ht_a( "$$site{rootp}/delete/$id", 'Delete' ),
								']' ),
						ht_utr() );
	}

	if ( db_rowcount( $sth ) < 1 ) {
		push( @lines, 	ht_tr(),
						ht_td( 	{ 'class' => 'cdta', 'colspan' => '3' }, 
								'No authors found.' ),
						ht_utr() );
	}

	db_finish( $sth );

	return( @lines, ht_utable(), ht_udiv() );
} # END $site->do_main

#-------------------------------------------------
# $site->do_photo( $r, $id )
#-------------------------------------------------
sub do_photo {
	my ( $site, $r, $id ) = @_;

	$$site{page_title}	.= 'Photo';

	return( 'Invalid id.' ) if ( ! is_integer( $id ) );

	my $apr    	= Apache2::Request->new( $r, TEMP_DIR => $$site{file_tmp} );
#                                            POST_MAX => 1023450 );
    #my $status  = $apr->body_status;
	#my $in = $apr->param;
	my $in = $site->param( $apr );
    
	#return( 'Error: Upload File too large.' ) if ( $status );

	if ( $in->{cancel} ) {
		return( $site->_relocate( $r, "$$site{rootp}/view/$id" ) );
	}

	my $sth = db_query( $$site{dbh}, 'get old values',
						'SELECT photo FROM bg_authors WHERE id = ',
						sql_num( $id ) );

	while ( my ( $photo ) = db_next( $sth ) ) {
		$in->{photo} = $photo 	if ( ! defined $in->{photo} );
	}

	db_finish( $sth );

	warn( 'A4' );
	if ( ! ( my @errors = photo_checkvals( $in, $apr ) ) ) {

		my $upload 			= $apr->upload( 'newfile' );
		my $type 			= $upload->type();
		$type 				=~ s/\//_/g;
		my ( $t, $fname )   = $upload->filename =~ /^(.*\\|.*\/)?(.*?)?$/;

		# Attach_path/$now-$uname-$type-$fname
		my $file = "$$site{file_path}/author_$id\_$fname";
		
		if ( open( ATTACH, ">$file" ) ) {

			my $fh = $upload->fh;

			while ( my $part = <$fh> ) {
				print ATTACH $part;
			}

			close( ATTACH );
		}
		else {
			die "Could not open $file: $!";
		}

		# Remove the old photo.
		if ( is_text( $in->{photo} ) ) {
			unlink( "$$site{file_path}/$in->{photo}" );
		}

		db_run( $$site{dbh}, 'insert new author',
				sql_update( 'bg_authors', 'WHERE id = '. sql_num( $id ),
							'photo' 	=> sql_str( "author_$id\_$fname" ) ) );

		db_commit( $$site{dbh} );

		return( $site->_relocate( $r, "$$site{rootp}/view/$id" ) );
	}
	else {
		if ( $r->method eq 'POST' ) {
			return( @errors, photo_form( $site, $in ) );
		}
		else {
			return( photo_form( $site, $in ) );
		}
	}
} # END $site->do_photo

#-------------------------------------------------
# $site->do_users( $r, $id )
#-------------------------------------------------
sub do_users {
	my ( $site, $r, $id ) = @_;

	return( 'Edit author users.' );
} # END $site->do_users

#-------------------------------------------------
# $site->do_view( $r, $id )
#-------------------------------------------------
sub do_view {
	my ( $site, $r, $id ) = @_;

	$$site{page_title} = 'Detail';

	return( 'Invalid id.' ) if ( ! is_integer( $id ) );

	my $sth = db_query( $$site{dbh}, 'look up the info',
						'SELECT ident, name, email, photo, bio',
						'FROM bg_authors WHERE id = ', sql_num( $id ) );
	
	my ( $ident, $name, $email, $photo, $bio ) = db_next( $sth );
	
	db_finish( $sth );

	my $image = '';
	my @photos;
	
	if ( $photo ) {
		$image = ht_img( "$$site{file_uri}/$photo", qq!alt="$name"! );
		push( @photos, ht_a( "$$site{rootp}/photo/$id", 'Update Photo' ), '|' );
	}
	else {
		push( @photos, ht_a( "$$site{rootp}/photo/$id", 'Add Photo' ), '|' );
	}


	return( '<h1>Preview</h1>',
			ht_div(),
			'[', ht_a( $$site{rootp}, 'Main' ), '|', @photos,
			ht_a( "$$site{rootp}/edit/$id", 'Edit' ), '|',
			ht_a( "$$site{rootp}/delete/$id", 'Delete'), ']',
			ht_udiv(),

			ht_div( { 'class' => 'author' } ),
			ht_div( { 'class' => 'author_header' } ),
			'<h1>', $name, '</h1>',
			'<h2>', $email, '</h2>',
			ht_udiv(),
			ht_div( { 'class' => 'author_content' } ),
			$image,
			$bio,
			ht_udiv(),
			ht_udiv() );
} # END $site->do_view

#-------------------------------------------------
# photo_checkvals( $in, $apr )
#-------------------------------------------------
sub photo_checkvals {
	my ( $in, $apr ) = @_;

	my @errors;

	if ( ! is_text( $in->{newfile} ) ) {
		push( @errors, 'Select a file to upload.'. ht_br() );
	}
	else { # Check the file size.
		my $upload 	= $apr->upload( 'newfile' );
		my $size 	= $upload->size;

		if ( ! defined $size || $size < 1 ) {
			push( @errors, 'Select a file to upload.'. ht_br() );
		}
	}

	return( @errors );
} # END photo_checkvals

#-------------------------------------------------
# photo_form( $site, $in )
#-------------------------------------------------
sub photo_form {
	my ( $site, $in ) = @_;

	my $photo = ($in->{photo}) ? ht_img( "$$site{file_uri}/$in->{photo}" ):
								'No photo';

	return( ht_form_js( $$site{uri}, 'enctype="multipart/form-data"' ),	
			ht_div( { 'class' => 'box' } ),
			ht_table( {} ),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'Current', ),
			ht_td( 	{ 'class' => 'dta' }, $photo ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'class' => 'shd' }, 'Upload', ),
			ht_td( 	{ 'class' => 'dta' },
					ht_input( 'newfile', 'file', $in ) ),
			ht_utr(),

			ht_tr(),
			ht_td( 	{ 'colspan' => '2', 'class' => 'rshd' }, 
					ht_submit( 'submit', 'Save' ),
					ht_submit( 'cancel', 'Cancel' ) ),
			ht_utr(),

			ht_utable(),
			ht_udiv(),
			ht_uform() );
} # END photo_form

#-------------------------------------------------
# users_checkvals( $in )
#-------------------------------------------------
sub users_checkvals {
	my ( $in ) = @_;

	my @errors = ( 'bad things' );

	return( @errors );
} # END users_checkvals()

#-------------------------------------------------
# users_form( $site, $in )
#-------------------------------------------------
sub users_form {
	my ( $site, $in ) = @_;

	return( 'users form' );
} # END users_form

# EOF
1;

__END__

=head1 NAME 

Alchemy::Blog::Author - Author management.

=head1 SYNOPSIS

  use Alchemy::Blog::Article::Author;

=head1 DESCRIPTION

This module manages authors that may be assigned to articles.

=head1 APACHE

This is a sample of the location required for this module to run.
Consult Alchemy::Blog(3) to learn about the configuration options.

  <Location /admin/authors >
    SetHandler  perl-script

    PerlHandler Alchemy::Blog::Author
  </Location>

=head1 DATABASE

This is the only table this module manipulates.

  create table "bg_authors" (
    id      int4 PRIMARY KEY DEFAULT NEXTVAL( 'bg_authors_seq' ),
    ident   varchar, 
    name    varchar,
    email   varchar,
    photo   varchar,
    bio     text
  );

=head1 SEE ALSO

Alchemy::Blog(3), Alchemy(3), KrKit(3)

=head1 LIMITATIONS

=head1 AUTHOR

Nicholas Studt <nstudt@angrydwarf.org>

=head1 COPYRIGHT

Copyright (c) 2003 Nicholas Studt. All rights reserved.

You may distribute under the terms of either the GNU General Public
License or the Artistic License, as specified in the Perl README file.

=cut
